### PR TITLE
FIX Ensure primary connection is updated with temporary database

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/11352

### CODE REVIEW

File: src/TestSessionEnvironment.php [hunk 1]
```diff
@@ -578,6 +578,10 @@ public function connectToDatabase($state = null)
                     $this->oldDatabaseName = $databaseConfig['database'];
                     $databaseConfig['database'] = $state->database;
                     DB::connect($databaseConfig);
+                    // This is required for CMS 5.4+
+                    if (method_exists(DB::class, 'hasConfig') && DB::hasConfig('primary')) {
+                        DB::connect($databaseConfig, 'primary');
+                    }
```
Comment (ChloeSartorelli) [inline src/TestSessionEnvironment.php]: We're connecting twice?
Shouldn't this be an if/else?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/TestSessionEnvironment.php (hunk 1 above)
Comment (emteknetnz) [reply src/TestSessionEnvironment.php]: Updated

File: src/TestSessionEnvironment.php [hunk 2]
```diff
@@ -613,4 +613,15 @@ public function waitForPendingRequests($await = 700, $timeout = 10000)
 
         return true;
     }
+
+    private function connectToDB(array $databaseConfig): void
+    {
+        if (method_exists(DB::class, 'hasConfig') && DB::hasConfig('primary')) {
+            // CMS 5.4+ - ensure we connect the primary connection and not a replica
+            DB::connect($databaseConfig, 'primary');
```
Comment (ChloeSartorelli) [inline src/TestSessionEnvironment.php]:
```suggestion
            DB::connect($databaseConfig, DB::CONN_PRIMARY);
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/TestSessionEnvironment.php (hunk 2 above)
Comment (ChloeSartorelli) [reply src/TestSessionEnvironment.php]: Also why are we checking `DB::hasConfig('primary')`? Don't we just care if the method exists (and therefore we're on 5.4+)?

Comment (emteknetnz) [reply src/TestSessionEnvironment.php]: Have removed the hasConfig() check, now just checking to see if the method exists

File: _config.php [hunk 3]
```diff
@@ -12,7 +12,7 @@
  */
 if (class_exists('Resque_Event') && class_exists('SSResqueRun')) {
     Resque_Event::listen('beforeFork', function ($data) {
-        $databaseConfig = DB::getConfig();
+        $databaseConfig = DB::getConfig(DB::CONN_PRIMARY);
```
Comment (ChloeSartorelli) [inline _config.php:15]: We'll need to bump the framework dependency to use that constant.
In that case we can remove the method check from `connectToDB` as well.

Comment (emteknetnz) [reply _config.php:15]: Done

File: src/TestSessionEnvironment.php [hunk 4]
```diff
@@ -618,4 +622,17 @@ public function waitForPendingRequests($await = 700, $timeout = 10000)
...
         return true;
     }
+
+    private function connectToDB(array $databaseConfig): void
+    {
+        if (method_exists(DB::class, 'hasConfig')) {
+            // CMS 5.4+ - ensure we connect the primary connection and not a replica
+            // which can happen if we use the default value of DB::CONN_DYNAMIC
+            // and there is a replica database configured
+            DB::connect($databaseConfig, DB::CONN_PRIMARY);
```
Comment (ChloeSartorelli) [inline src/TestSessionEnvironment.php]: This PR targets CMS 6 now. We don't need this condition (and therefore probably don't need this method at all)

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/TestSessionEnvironment.php (hunk 4 above)
Comment (emteknetnz) [reply src/TestSessionEnvironment.php]: Updated

**Review by GuySartorelli — APPROVED** [review]
LGTM, will merge when other PRs are all approved

**Pull request merged**